# Practico Mentoria - Analisis Exploratorio y Curación de Datos

---

### Importaciones

In [8]:
%matplotlib inline

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import scipy as sp

from sklearn import preprocessing

import warnings
warnings.filterwarnings('ignore')

In [9]:
# Set seed for reproducibility
np.random.seed(0)

In [10]:
player_df = pd.read_csv('./Datasets/football_player.csv')
team_df = pd.read_csv('./Datasets/football_team.csv')
match_df = pd.read_csv('./Datasets/football_match.csv')

In [11]:
print("Shape 'player_df' = {}".format(player_df.shape))
print("Shape 'team_df'   = {}".format(team_df.shape))
print("Shape 'match_df'  = {}".format(match_df.shape))

Shape 'player_df' = (11060, 40)
Shape 'team_df'   = (288, 22)
Shape 'match_df'  = (25979, 12)


---

# 1. Importacion de los datos

Calculemos el rango de fechas de los partidos

In [12]:
match_df['date'].max() - match_df['date'].min()

TypeError: unsupported operand type(s) for -: 'str' and 'str'

Indiquemos que la columna `date` es una fechas como indica la [documentación](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#datetime-handling)

In [19]:
match_df['date'].max()

'2016-05-25 00:00:00'

In [21]:
match_df['date'].min()

'2008-07-18 00:00:00'

In [23]:
match_df.dtypes

country name            object
league name             object
season                  object
stage                    int64
date                    object
home team long name     object
home short long name    object
away team long name     object
away short long name    object
home team goal           int64
away team goal           int64
total goal               int64
dtype: object

In [26]:
match_df['Fecha'] =  pd.to_datetime(match_df['date'])


In [27]:
match_df.dtypes

country name                    object
league name                     object
season                          object
stage                            int64
date                            object
home team long name             object
home short long name            object
away team long name             object
away short long name            object
home team goal                   int64
away team goal                   int64
total goal                       int64
Fecha                   datetime64[ns]
dtype: object

In [32]:
match_df['Fecha']

0       2008-08-17
1       2008-08-16
2       2008-08-16
3       2008-08-17
4       2008-08-16
5       2008-09-24
6       2008-08-16
7       2008-08-16
8       2008-08-16
9       2008-11-01
10      2008-10-31
11      2008-11-02
12      2008-11-01
13      2008-11-01
14      2008-11-01
15      2008-11-01
16      2008-11-01
17      2008-11-02
18      2008-11-08
19      2008-11-08
20      2008-11-09
21      2008-11-07
22      2008-11-08
23      2008-11-08
24      2008-11-08
25      2008-11-08
26      2008-11-09
27      2008-11-16
28      2008-11-15
29      2008-11-15
           ...    
25949   2015-08-08
25950   2015-08-08
25951   2015-08-09
25952   2015-08-09
25953   2015-08-09
25954   2015-08-12
25955   2015-08-12
25956   2015-08-12
25957   2015-08-13
25958   2015-08-13
25959   2015-08-22
25960   2015-08-23
25961   2015-08-23
25962   2015-08-23
25963   2015-08-22
25964   2015-08-29
25965   2015-08-29
25966   2015-08-30
25967   2015-08-30
25968   2015-08-30
25969   2015-09-12
25970   2015

In [33]:
match_df['Fecha'].max()

Timestamp('2016-05-25 00:00:00')

In [34]:
match_df['Fecha'].min()

Timestamp('2008-07-18 00:00:00')

In [35]:
match_df['Fecha'].max() - match_df['Fecha'].min()

Timedelta('2868 days 00:00:00')

Existen 2868 días de diferencia entre la fecha minima y la maxima

Por otro lado podríamos leer directamente el dataframe paresando en la lectura el campo DATE a fecha

In [40]:
match_df_parseDate = pd.read_csv('./Datasets/football_match.csv', parse_dates=["date"])
match_df_parseDate.dtypes

country name                    object
league name                     object
season                          object
stage                            int64
date                    datetime64[ns]
home team long name             object
home short long name            object
away team long name             object
away short long name            object
home team goal                   int64
away team goal                   int64
total goal                       int64
dtype: object

In [41]:
match_df_parseDate['date'].max() - match_df_parseDate['date'].min()

Timedelta('2868 days 00:00:00')

Es otra forma de hacer el calculo

# 2. Etiquetas de variables/columnas: no usar caracteres especiales

Chequar que no haya caracteres fuera de `a-Z`, `0-9` y `_` en los nombres de columnas de los Dataframes:
* `player_df`
* `team_df`
* `match_df`

In [36]:
player_df.describe()

height_m     weight_kg  overall_rating     potential      crossing  \
count  11060.000000  11060.000000    11060.000000  11060.000000  11060.000000   
mean       1.817847     76.375393       66.821222     72.090216     52.853855   
std        0.063278      6.799564        6.237719      5.800313     16.169989   
min        1.570000     53.070000       43.000000     51.000000      6.000000   
25%        1.780000     72.120000       62.820000     68.000000     43.440000   
50%        1.830000     76.200000       66.720000     72.000000     56.300000   
75%        1.850000     81.190000       70.952500     76.000000     64.710000   
max        2.080000    110.220000       92.190000     95.230000     89.360000   

          finishing  heading accuracy  short passing       volleys  \
count  11060.000000      11060.000000   11060.000000  10582.000000   
mean      47.862155         56.100191      60.367143     47.110973   
std       18.109552         15.655413      13.508685     17.340289   
min        5.000000          8.000000      10.570000      3.750000   
25%       32.440000         49.097500      55.620000     33.250000   
50%       49.855000         58.805000      63.000000     49.300000   
75%       63.060000         66.750000      69.007500     60.737500   
max       92.230000         93.110000      95.180000     90.790000   

          dribbling  ...        vision     penalties       marking  \
count  11060.000000  ...  10582.000000  11060.000000  11060.000000   
mean      56.892354  ...     55.866959     53.683902     46.231480   
std       16.861731  ...     14.046464     13.837462     20.055209   
min        5.140000  ...      8.000000      9.000000      5.000000   
25%       49.110000  ...     47.330000     44.850000     25.000000   
50%       61.100000  ...     57.730000     55.000000     50.000000   
75%       68.782500  ...     66.000000     63.930000     64.200000   
max       96.460000  ...     95.680000     92.000000     90.670000   

       standing tackle  sliding tackle     gk_diving   gk_handling  \
count     11060.000000    10582.000000  11060.000000  11060.000000   
mean         49.440716       47.124626     14.911413     16.676659   
std          20.336005       20.591114     16.804792     15.472431   
min           6.000000        5.000000      1.000000      2.000000   
25%          29.000000       25.440000      7.770000      9.500000   
50%          55.190000       52.165000     10.140000     12.500000   
75%          67.000000       65.120000     13.000000     15.390000   
max          90.200000       94.370000     89.860000     86.880000   

         gk_kicking  gk_positioning   gk_reflexes  
count  11060.000000    11060.000000  11060.000000  
mean      23.015325       16.750476     17.055242  
std       17.583860       15.692518     16.754717  
min        3.260000        2.000000      2.000000  
25%       11.000000        9.437500      9.380000  
50%       16.280000       12.500000     12.460000  
75%       27.302500       15.422500     15.380000  
max       89.000000       91.620000     90.950000  

[8 rows x 37 columns]

In [37]:
match_df.describe()

stage  home team goal  away team goal    total goal
count  25979.000000    25979.000000    25979.000000  25979.000000
mean      18.242773        1.544594        1.160938      2.705531
std       10.407354        1.297158        1.142110      1.672456
min        1.000000        0.000000        0.000000      0.000000
25%        9.000000        1.000000        0.000000      2.000000
50%       18.000000        1.000000        1.000000      3.000000
75%       27.000000        2.000000        2.000000      4.000000
max       38.000000       10.000000        9.000000     12.000000

In [38]:
team_df.describe()

buildUpPlaySpeed  buildUpPlayPassing  chanceCreationPassing  \
count        288.000000          288.000000             288.000000   
mean          52.026458           48.725000              51.726736   
std            7.742841            7.351824               6.608335   
min           29.330000           30.000000              30.000000   
25%           47.297500           44.000000              48.000000   
50%           52.330000           48.450000              51.450000   
75%           57.347500           53.170000              55.702500   
max           72.500000           77.000000              68.830000   

       chanceCreationCrossing  chanceCreationShooting  defencePressure  \
count              288.000000              288.000000       288.000000   
mean                53.401493               53.884792        45.821250   
std                  7.544098                6.239400         6.999462   
min                 33.330000               36.670000        25.670000   
25%                 49.297500               50.200000        41.297500   
50%                 53.635000               53.915000        46.000000   
75%                 58.542500               58.000000        50.042500   
max                 77.000000               70.000000        70.000000   

       defenceAggression  defenceTeamWidth  
count         288.000000        288.000000  
mean           49.153403         52.044167  
std             6.186604          6.421537  
min            30.000000         30.000000  
25%            45.600000         48.230000  
50%            48.735000         52.450000  
75%            52.830000         55.872500  
max            70.000000         70.000000

In [42]:
import string

string.ascii_letters + string.digits

'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'

Estos son los caracteres normales, de los cuales ahora vamos a quitar los caracteres de las columnas de los datasets que no entran en ese conjunto

In [43]:
team_df.columns[~team_df.columns.str.match(r'^(\w+)$')]

Index(['team long name', 'team short name'], dtype='object')

team long name y team short name tienen nombres con espacios. Los vamos a quitar reemplazandolos con algun valor. 

In [44]:
team_df.rename(columns={'team long name':'teamLongName', 'team short name':'teamShortName', }, inplace=True)

In [45]:
team_df.columns

Index(['teamLongName', 'teamShortName', 'buildUpPlaySpeed',
       'buildUpPlaySpeedClass', 'buildUpPlayDribblingClass',
       'buildUpPlayPassing', 'buildUpPlayPassingClass',
       'buildUpPlayPositioningClass', 'chanceCreationPassing',
       'chanceCreationPassingClass', 'chanceCreationCrossing',
       'chanceCreationCrossingClass', 'chanceCreationShooting',
       'chanceCreationShootingClass', 'chanceCreationPositioningClass',
       'defencePressure', 'defencePressureClass', 'defenceAggression',
       'defenceAggressionClass', 'defenceTeamWidth', 'defenceTeamWidthClass',
       'defenceDefenderLineClass'],
      dtype='object')

In [48]:
match_df.columns[~match_df.columns.str.match(r'^(\w+)$')]

Index(['country name', 'league name', 'home team long name',
       'home short long name', 'away team long name', 'away short long name',
       'home team goal', 'away team goal', 'total goal'],
      dtype='object')

In [49]:
match_df.columns = match_df.columns.str.replace(' ', '_')
match_df.columns

Index(['country_name', 'league_name', 'season', 'stage', 'date',
       'home_team_long_name', 'home_short_long_name', 'away_team_long_name',
       'away_short_long_name', 'home_team_goal', 'away_team_goal',
       'total_goal', 'Fecha'],
      dtype='object')

Reemplazamos los espacios en esta oportunidad con '_'

In [47]:
player_df.columns[~player_df.columns.str.match(r'^(\w+)$')]

Index(['player name', 'preferred foot', 'heading accuracy', 'short passing',
       'free kick accuracy', 'long passing', 'ball control', 'sprint speed',
       'shot power', 'long shots', 'standing tackle', 'sliding tackle'],
      dtype='object')

In [50]:
player_df.columns = player_df.columns.str.replace(' ', '_')
player_df.columns

Index(['player_name', 'birthday', 'height_m', 'weight_kg', 'overall_rating',
       'potential', 'preferred_foot', 'crossing', 'finishing',
       'heading_accuracy', 'short_passing', 'volleys', 'dribbling', 'curve',
       'free_kick_accuracy', 'long_passing', 'ball_control', 'acceleration',
       'sprint_speed', 'agility', 'reactions', 'balance', 'shot_power',
       'jumping', 'stamina', 'strength', 'long_shots', 'aggression',
       'interceptions', 'positioning', 'vision', 'penalties', 'marking',
       'standing_tackle', 'sliding_tackle', 'gk_diving', 'gk_handling',
       'gk_kicking', 'gk_positioning', 'gk_reflexes'],
      dtype='object')

# 3. Agregar nuevas caracteristicas

Agregar al Dataframe `player_df` una nueva columna que sea `imc` correspondiente al **Indice de Masa Corporal**

Link:
* https://www.texasheart.org/heart-health/heart-information-center/topics/calculadora-del-indice-de-masa-corporal-imc/

In [67]:
def imc(peso,altura):
    return peso / (altura**2)

player_df['IMC'] = player_df.apply(lambda x: imc(x.weight_kg, x.height_m), axis=1)


In [68]:
player_df[['height_m','weight_kg','IMC']]

height_m  weight_kg        IMC
0          1.83      84.82  25.327720
1          1.70      66.22  22.913495
2          1.70      73.94  25.584775
3          1.83      89.81  26.817761
4          1.83      69.85  20.857595
5          1.83      73.03  21.807161
6          1.73      66.22  22.125697
7          1.65      63.05  23.158861
8          1.90      82.10  22.742382
9          1.75      77.11  25.178776
10         1.83      82.10  24.515513
11         1.75      68.04  22.217143
12         1.88      76.20  21.559529
13         1.70      64.86  22.442907
14         1.78      69.85  22.045828
15         1.73      73.94  24.705135
16         1.83      79.83  23.837678
17         1.90      72.12  19.977839
18         1.75      68.04  22.217143
19         1.88      83.91  23.740946
20         1.78      73.03  23.049489
21         1.85      73.03  21.338203
22         1.80      73.03  22.540123
23         1.83      78.02  23.297202
24         1.80      73.03  22.540123
25         1.73      64.86  21.671289
26         1.85      74.84  21.867056
27         1.80      72.12  22.259259
28         1.83      77.11  23.025471
29         1.78      71.21  22.475066
...         ...        ...        ...
11030      1.75      73.03  23.846531
11031      1.80      74.84  23.098765
11032      1.75      66.22  21.622857
11033      1.78      64.86  20.470900
11034      1.80      77.11  23.799383
11035      1.73      69.85  23.338568
11036      1.85      73.94  21.604091
11037      1.98      92.99  23.719518
11038      1.80      73.94  22.820988
11039      1.75      67.13  21.920000
11040      1.83      78.93  23.568933
11041      1.96      94.80  24.677218
11042      1.85      79.83  23.325055
11043      1.83      77.11  23.025471
11044      1.88      83.01  23.486306
11045      1.73      68.95  23.037856
11046      1.65      64.86  23.823691
11047      1.83      74.84  22.347637
11048      1.75      67.13  21.920000
11049      1.83      79.83  23.837678
11050      1.88      74.84  21.174740
11051      1.90      81.19  22.490305
11052      1.70      71.21  24.640138
11053      1.80      72.12  22.259259
11054      1.90      78.02  21.612188
11055      1.83      76.20  22.753740
11056      1.83      79.83  23.837678
11057      1.80      69.85  21.558642
11058      1.85      78.02  22.796202
11059      1.80      79.83  24.638889

[11060 rows x 3 columns]

In [66]:
player_df.describe()

height_m     weight_kg  overall_rating     potential      crossing  \
count  11060.000000  11060.000000    11060.000000  11060.000000  11060.000000   
mean       1.817847     76.375393       66.821222     72.090216     52.853855   
std        0.063278      6.799564        6.237719      5.800313     16.169989   
min        1.570000     53.070000       43.000000     51.000000      6.000000   
25%        1.780000     72.120000       62.820000     68.000000     43.440000   
50%        1.830000     76.200000       66.720000     72.000000     56.300000   
75%        1.850000     81.190000       70.952500     76.000000     64.710000   
max        2.080000    110.220000       92.190000     95.230000     89.360000   

          finishing  heading_accuracy  short_passing       volleys  \
count  11060.000000      11060.000000   11060.000000  10582.000000   
mean      47.862155         56.100191      60.367143     47.110973   
std       18.109552         15.655413      13.508685     17.340289   
min        5.000000          8.000000      10.570000      3.750000   
25%       32.440000         49.097500      55.620000     33.250000   
50%       49.855000         58.805000      63.000000     49.300000   
75%       63.060000         66.750000      69.007500     60.737500   
max       92.230000         93.110000      95.180000     90.790000   

          dribbling  ...       marking  standing_tackle  sliding_tackle  \
count  11060.000000  ...  11060.000000     11060.000000    10582.000000   
mean      56.892354  ...     46.231480        49.440716       47.124626   
std       16.861731  ...     20.055209        20.336005       20.591114   
min        5.140000  ...      5.000000         6.000000        5.000000   
25%       49.110000  ...     25.000000        29.000000       25.440000   
50%       61.100000  ...     50.000000        55.190000       52.165000   
75%       68.782500  ...     64.200000        67.000000       65.120000   
max       96.460000  ...     90.670000        90.200000       94.370000   

          gk_diving   gk_handling    gk_kicking  gk_positioning   gk_reflexes  \
count  11060.000000  11060.000000  11060.000000    11060.000000  11060.000000   
mean      14.911413     16.676659     23.015325       16.750476     17.055242   
std       16.804792     15.472431     17.583860       15.692518     16.754717   
min        1.000000      2.000000      3.260000        2.000000      2.000000   
25%        7.770000      9.500000     11.000000        9.437500      9.380000   
50%       10.140000     12.500000     16.280000       12.500000     12.460000   
75%       13.000000     15.390000     27.302500       15.422500     15.380000   
max       89.860000     86.880000     89.000000       91.620000     90.950000   

              col_3           IMC  
count  11060.000000  11060.000000  
mean     139.168414     23.086261  
std       16.403700      1.319189  
min       83.319900     17.581485  
25%      127.825500     22.259259  
50%      138.798000     23.062089  
75%      150.080400     23.872576  
max      229.257600     30.766147  

[8 rows x 39 columns]

# 4. Tratar valores faltantes

Veamos cuantos valores nulos tenemos

In [69]:
player_missing_values_count = player_df.isnull().sum()

player_missing_values_count[player_missing_values_count > 0]

volleys           478
curve             478
agility           478
balance           478
jumping           478
vision            478
sliding_tackle    478
dtype: int64

In [70]:
team_missing_values_count = team_df.isnull().sum()

team_missing_values_count[team_missing_values_count > 0]

Series([], dtype: int64)

In [71]:
match_missing_values_count = match_df.isnull().sum()

match_missing_values_count[match_missing_values_count > 0]

Series([], dtype: int64)

Algunas tecnicas para tratar los _missing values_:
* **Eliminar** muestras o variables que tienen datos faltantes.
* **Imputar** los valores perdidos, es decir, sustituirlos por estimaciones por ejemplo la `media`, la `moda` ó usando `KNN`.

A) Analizar si es conveniente **Eliminar** las muestras o variables con datos faltantes del Dataframe `player_df`. 

B) Aplicar la **Imputacion** usando la `media` o `moda` sobre las columnas con _missing values_ del Dataframe `player_df`.

**Hint**:
* Para la imputacion usando la `media`, `moda` ver el siguiente link:  
    https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html

### ¿Eliminar los _missing values_? Justificar

In [72]:
len(player_df.dropna())/len(player_df)

0.9567811934900543

In [76]:
len(player_df.dropna(subset=['volleys']))/len(player_df)

0.9567811934900543

In [80]:
player_df.shape

(11060, 42)

In [83]:
player_df[player_df.volleys.isnull()]

player_name    birthday  height_m  weight_kg  \
25            Abdelmajid Oulmers  1978-09-12      1.73      64.86   
30              Abdeslam Ouaddou  1978-11-01      1.90      82.10   
31         Abdessalam Benjelloun  1985-01-28      1.88      81.19   
83                   Aco Stojkov  1983-04-29      1.78      74.84   
85                      Adailton  1977-01-24      1.75      73.03   
175           Adrian Paluchowski  1987-08-19      1.80      74.84   
190                      Adriano  1982-01-21      1.75      76.20   
203              Afonso Alves,24  1981-01-30      1.85      73.94   
253                 Alan Haydock  1976-01-13      1.75      72.12   
275                Albert Baning  1985-03-19      1.93      81.19   
289              Alberto Fontana  1967-01-23      1.85      73.03   
343          Aleksandar Mitreski  1980-08-05      1.85      74.84   
379          Alessandro Grandoni  1977-07-22      1.80      76.20   
405                   Alex Bruno  1982-05-09      1.88      86.18   
456               Alexander Laas  1984-05-05      1.73      69.85   
476        Alexandre Di Gregorio  1980-02-12      1.75      79.83   
485            Alexandre Quennoz  1978-09-21      1.80      79.83   
534                Allan Russell  1980-12-13      1.85      77.11   
577              Amadou Alassane  1983-04-07      1.88      76.20   
584                    Amdy Faye  1977-03-12      1.83      78.02   
655                  Andre Leone  1979-02-12      1.83      81.19   
675                Andrea Ardito  1977-01-08      1.73      60.78   
713             Andrea Zanchetta  1975-02-02      1.80      73.94   
756                Andrew McNeil  1987-01-19      1.80      83.01   
774            Andriy Shevchenko  1976-09-29      1.83      72.12   
798    Angel Manuel Vivar Dorado  1974-02-12      1.83      78.02   
804                Angelo Martha  1982-04-29      1.85      82.10   
827               Anthony Bentem  1990-03-19      1.80      74.84   
885    Antonio Carlos Dos Santos  1979-10-03      1.88      68.04   
887             Antonio Chimenti  1970-06-30      1.83      83.01   
...                          ...         ...       ...        ...   
10449             Tony Heurtebis  1975-01-15      1.80      73.03   
10475           Tugay Kerimoglou  1970-08-24      1.75      73.03   
10499              Ulrich Le-Pen  1974-01-23      1.75      68.04   
10504                  Umit Ozat  1976-10-30      1.85      73.94   
10528            Vahid Hashemian  1976-07-21      1.83      78.02   
10545            Valerien Ismael  1975-09-28      1.90      83.01   
10575         Veldin Muharemovic  1984-12-06      1.83      77.11   
10627             Vincent Hognon  1974-08-16      1.83      74.84   
10640           Vincent Provoost  1984-02-07      1.78      76.20   
10670                 Vitor Lima  1981-08-18      1.78      63.96   
10671               Vitor Moreno  1980-11-29      1.80      86.18   
10674           Vittorio Villano  1988-02-02      1.65      62.14   
10684         Vladimir Stojkovic  1983-07-28      1.93      93.89   
10692       Vojtech Schulmeister  1983-09-09      1.83      78.93   
10696              Vukasin Devic  1984-03-15      1.85      74.84   
10711             Walid Regragui  1975-09-23      1.78      66.22   
10738                 Wellington  1985-08-17      1.75      71.21   
10742                     Wender  1975-04-17      1.78      73.94   
10757                    Wesllem  1985-04-21      1.75      72.12   
10802              Willy Grondin  1974-10-12      1.78      78.02   
10894               Yasin Karaca  1983-12-16      1.70      67.13   
10909          Yazid Mansouri,30  1978-02-25      1.75      68.95   
10913           Yildiray Basturk  1978-12-24      1.70      68.95   
10924                   Yoav Ziv  1981-03-16      1.75      74.84   
10934            Yohan Lachor,29  1976-08-03      1.83      79.83   
10949              Yoshito Okubo  1982-06-09      1.68      60.78   
10978            Yu

In [94]:
player_df_nulos = player_df[player_df.volleys.isnull()]
player_df_nulos[['volleys','agility','curve','balance','jumping','vision','sliding_tackle']]

volleys  agility  curve  balance  jumping  vision  sliding_tackle
25         NaN      NaN    NaN      NaN      NaN     NaN             NaN
30         NaN      NaN    NaN      NaN      NaN     NaN             NaN
31         NaN      NaN    NaN      NaN      NaN     NaN             NaN
83         NaN      NaN    NaN      NaN      NaN     NaN             NaN
85         NaN      NaN    NaN      NaN      NaN     NaN             NaN
175        NaN      NaN    NaN      NaN      NaN     NaN             NaN
190        NaN      NaN    NaN      NaN      NaN     NaN             NaN
203        NaN      NaN    NaN      NaN      NaN     NaN             NaN
253        NaN      NaN    NaN      NaN      NaN     NaN             NaN
275        NaN      NaN    NaN      NaN      NaN     NaN             NaN
289        NaN      NaN    NaN      NaN      NaN     NaN             NaN
343        NaN      NaN    NaN      NaN      NaN     NaN             NaN
379        NaN      NaN    NaN      NaN      NaN     NaN             NaN
405        NaN      NaN    NaN      NaN      NaN     NaN             NaN
456        NaN      NaN    NaN      NaN      NaN     NaN             NaN
476        NaN      NaN    NaN      NaN      NaN     NaN             NaN
485        NaN      NaN    NaN      NaN      NaN     NaN             NaN
534        NaN      NaN    NaN      NaN      NaN     NaN             NaN
577        NaN      NaN    NaN      NaN      NaN     NaN             NaN
584        NaN      NaN    NaN      NaN      NaN     NaN             NaN
655        NaN      NaN    NaN      NaN      NaN     NaN             NaN
675        NaN      NaN    NaN      NaN      NaN     NaN             NaN
713        NaN      NaN    NaN      NaN      NaN     NaN             NaN
756        NaN      NaN    NaN      NaN      NaN     NaN             NaN
774        NaN      NaN    NaN      NaN      NaN     NaN             NaN
798        NaN      NaN    NaN      NaN      NaN     NaN             NaN
804        NaN      NaN    NaN      NaN      NaN     NaN             NaN
827        NaN      NaN    NaN      NaN      NaN     NaN             NaN
885        NaN      NaN    NaN      NaN      NaN     NaN             NaN
887        NaN      NaN    NaN      NaN      NaN     NaN             NaN
...        ...      ...    ...      ...      ...     ...             ...
10449      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10475      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10499      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10504      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10528      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10545      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10575      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10627      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10640      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10670      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10671      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10674      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10684      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10692      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10696      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10711      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10738      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10742      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10757      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10802      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10894      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10909      NaN      NaN    NaN      NaN      NaN     NaN             NaN
10913      NaN      NaN    NaN      NaN      NaN     NaN         

In [95]:
player_df_nulos[['volleys','agility','curve','balance','jumping','vision','sliding_tackle']].dtypes

volleys           float64
agility           float64
curve             float64
balance           float64
jumping           float64
vision            float64
sliding_tackle    float64
dtype: object

Como podremos observar de los 478 NaN q se encuentran en el dataset, todas las columnas presentan el mismo valor para los mismos jugadores. Es posible de eliminar dichos jugadores que no contengan esta información ya que tenemos 11060 muestras para continuar con nuestro analisis. 

In [89]:
player_df_sinNulos = player_df.dropna()
player_df_sinNulos.shape

(10582, 42)

In [91]:
player_df.dtypes

player_name            object
birthday               object
height_m              float64
weight_kg             float64
overall_rating        float64
potential             float64
preferred_foot         object
crossing              float64
finishing             float64
heading_accuracy      float64
short_passing         float64
volleys               float64
dribbling             float64
curve                 float64
free_kick_accuracy    float64
long_passing          float64
ball_control          float64
acceleration          float64
sprint_speed          float64
agility               float64
reactions             float64
balance               float64
shot_power            float64
jumping               float64
stamina               float64
strength              float64
long_shots            float64
aggression            float64
interceptions         float64
positioning           float64
vision                float64
penalties             float64
marking               float64
standing_t

In [92]:
del player_df['col_3']
del player_df_sinNulos['col_3']

In [93]:
player_df.dtypes

player_name            object
birthday               object
height_m              float64
weight_kg             float64
overall_rating        float64
potential             float64
preferred_foot         object
crossing              float64
finishing             float64
heading_accuracy      float64
short_passing         float64
volleys               float64
dribbling             float64
curve                 float64
free_kick_accuracy    float64
long_passing          float64
ball_control          float64
acceleration          float64
sprint_speed          float64
agility               float64
reactions             float64
balance               float64
shot_power            float64
jumping               float64
stamina               float64
strength              float64
long_shots            float64
aggression            float64
interceptions         float64
positioning           float64
vision                float64
penalties             float64
marking               float64
standing_t

### Imputacion usando Media y Moda

In [97]:
player_df[['volleys','agility','curve','balance','jumping','vision','sliding_tackle']].describe()

volleys       agility         curve       balance       jumping  \
count  10582.000000  10582.000000  10582.000000  10582.000000  10582.000000   
mean      47.110973     64.409834     50.423617     64.475489     66.038396   
std       17.340289     12.166882     17.309270     11.262069      9.555309   
min        3.750000     21.000000      5.710000     20.000000     21.000000   
25%       33.250000     57.000000     38.760000     58.160000     60.000000   
50%       49.300000     65.730000     52.565000     65.670000     66.890000   
75%       60.737500     73.000000     63.570000     72.000000     72.360000   
max       90.790000     94.670000     92.570000     94.310000     94.310000   

             vision  sliding_tackle  
count  10582.000000    10582.000000  
mean      55.866959       47.124626  
std       14.046464       20.591114  
min        8.000000        5.000000  
25%       47.330000       25.440000  
50%       57.730000       52.165000  
75%       66.000000       65.120000  
max       95.680000       94.370000

In [99]:
player_df.mode()

player_name    birthday  height_m  weight_kg  overall_rating  potential  \
0      Danilo  1989-03-02      1.83      74.84            64.0       67.0   

  preferred_foot  crossing  finishing  heading_accuracy  ...  penalties  \
0          right      49.0       25.0              52.0  ...       45.0   

   marking  standing_tackle  sliding_tackle  gk_diving  gk_handling  \
0     25.0             25.0            25.0        9.0         14.0   

   gk_kicking  gk_positioning  gk_reflexes        IMC  
0         9.0            11.0         12.0  22.347637  

[1 rows x 41 columns]

In [107]:
player_df_reemplazo_nan_moda = player_df

for column in ['volleys','agility','curve','balance','jumping','vision','sliding_tackle']:
    player_df_reemplazo_nan_moda[column].fillna(player_df_reemplazo_nan_moda[column].mode()[0], inplace=True)

player_df_reemplazo_nan_moda

player_name    birthday  height_m  weight_kg  \
0          Aaron Appindangoye  1992-02-29      1.83      84.82   
1             Aaron Cresswell  1989-12-15      1.70      66.22   
2                 Aaron Doran  1991-05-13      1.70      73.94   
3               Aaron Galindo  1982-05-08      1.83      89.81   
4                Aaron Hughes  1979-11-08      1.83      69.85   
5                  Aaron Hunt  1986-09-04      1.83      73.03   
6                  Aaron Kuhl  1996-01-30      1.73      66.22   
7                Aaron Lennon  1987-04-16      1.65      63.05   
8                Aaron Lennox  1993-02-19      1.90      82.10   
9               Aaron Meijers  1987-10-28      1.75      77.11   
10              Aaron Mokoena  1980-11-25      1.83      82.10   
11                 Aaron Mooy  1990-09-15      1.75      68.04   
12             Aaron Muirhead  1990-08-30      1.88      76.20   
13               Aaron Niguez  1989-04-26      1.70      64.86   
14               Aaron Ramsey  1990-12-26      1.78      69.85   
15              Aaron Splaine  1996-10-13      1.73      73.94   
16      Aaron Taylor-Sinclair  1991-04-08      1.83      79.83   
17            Aaron Wilbraham  1979-10-21      1.90      72.12   
18          Aatif Chahechouhe  1986-07-02      1.75      68.04   
19                  Abasse Ba  1976-07-12      1.88      83.91   
20          Abdelaziz Barrada  1989-06-19      1.78      73.03   
21      Abdelfettah Boukhriss  1986-10-22      1.85      73.03   
22     Abdelhamid El Kaoutari  1990-03-17      1.80      73.03   
23         Abdelkader Ghezzal  1984-12-05      1.83      78.02   
24            Abdellah Zoubir  1991-12-05      1.80      73.03   
25         Abdelmajid Oulmers  1978-09-12      1.73      64.86   
26         Abdelmalek Cherrad  1981-01-14      1.85      74.84   
27     Abdelmalek El Hasnaoui  1994-02-09      1.80      72.12   
28        Abdelouahed Chakhsi  1986-10-01      1.83      77.11   
29           Abderrazak Jadid  1983-06-01      1.78      71.21   
...                       ...         ...       ...        ...   
11030                 Zezinho  1992-09-23      1.75      73.03   
11031               Zhi Zheng  1980-08-20      1.80      74.84   
11032             Zhi-Gin Lam  1991-06-04      1.75      66.22   
11033             Zie Diabate  1989-03-02      1.78      64.86   
11034            Ziggy Gordon  1993-04-23      1.80      77.11   
11035         Ziguy Badibanga  1991-11-26      1.73      69.85   
11036        Zinedine Machach  1996-01-05      1.85      73.94   
11037              Zinho Gano  1993-10-13      1.98      92.99   
11038             Ziri Hammar  1992-07-25      1.80      73.94   
11039                    Zizo  1996-01-10      1.75      67.13   
11040       Zlatan Bajramovic  1979-08-12      1.83      78.93   
11041      Zlatan Ibrahimovic  1981-10-03      1.96      94.80   
11042      Zlatan Ljubijankic  1983-12-15      1.85      79.83   
11043            Zlatko Dedic  1984-10-05      1.83      77.11   
11044           Zlatko Janjic  1986-05-07      1.88      83.01   
11045        Zlatko Junuzovic  1987-09-26      1.73      68.95   
11046           Zola Matumona  1981-11-26      1.65      64.86   
11047             Zoltan Gera  1979-04-22      1.83      74.84   
11048          Zoltan Stieber  1988-10-16      1.75      67.13   
11049          Zoltan Szelesi  1981-11-22      1.83      79.83   
11050         Zoran Josipovic  1995-08-25      1.88      74.84   
11051          Zoran Rendulic  1984-05-22      1.90      81.19   
11052             Zoran Tosic  1987-04-28      1.70      71.21   
11053      Zouhaier Dhaouadhi  1988-01-01      1.80      72.12   
11054          Zouhair Feddal  1989-01-01      1.90      78.02   
11055          Zoumana Camara  1979-04-03      1.83      76.20   
11056            Zsolt Laczko  1986-12-18      1.83      79.83   
11057               Zsolt Low  1979-04-29      1.80      69.85   
11058     Zurab Khizanishvili  1981-10-06      1.85      7

In [109]:
player_df_reemplazo_nan_moda

player_missing_values_count = player_df_reemplazo_nan_moda.isnull().sum()

player_missing_values_count[player_missing_values_count > 0]

Series([], dtype: int64)

NO se puede calcular con moda porque los valores son decimales (float). No es entero el numero

# 5. Normalizacion de columnas

Primero que todo la notación:
* $x = [x_1, x_2, ..., x_n]$
* $\mu$: Media
* $\sigma$: Desviacion Estandar

Ahora normalizaremos algunas de las columnas del Dataframe, para ello usaremos dos tipos de normalizacion:

* Min-Max:  
$$
z_i = \frac{x_i - min(x)}{max(x) - min(x)}
$$
* Z-score
$$
z_i = \frac{x_i - \mu}{\sigma}
$$

Normalizar la columna `crossing` usando **Min-Max**.

Normalizar la columna `short_passing` usando **Z-score**.

**Hints**:
* https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler
* https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.zscore.html

### Min-Max

In [113]:
import pandas as pd
from sklearn import preprocessing

In [115]:
scaler = preprocessing.MinMaxScaler()
player_df[["crossing"]] = scaler.fit_transform(player_df[["crossing"]])
player_df[["crossing"]]

crossing
0      0.511036
1      0.777231
2      0.745202
3      0.614443
4      0.468810
5      0.814419
6      0.498680
7      0.864203
8      0.071977
9      0.694458
10     0.239923
11     0.726607
12     0.503839
13     0.642514
14     0.802303
15     0.580374
16     0.643114
17     0.496161
18     0.761756
19     0.419866
20     0.770274
21     0.491843
22     0.661468
23     0.689779
24     0.515835
25     0.647793
26     0.515835
27     0.431862
28     0.563820
29     0.704055
...         ...
11030  0.563820
11031  0.767754
11032  0.727087
11033  0.585773
11034  0.620202
11035  0.647793
11036  0.697097
11037  0.514995
11038  0.731766
11039  0.419866
11040  0.814060
11041  0.796305
11042  0.753359
11043  0.617802
11044  0.687500
11045  0.812620
11046  0.704295
11047  0.910749
11048  0.699376
11049  0.622121
11050  0.263916
11051  0.371881
11052  0.755758
11053  0.707774
11054  0.498081
11055  0.431862
11056  0.734765
11057  0.685461
11058  0.488844
11059  0.866123

[11060 rows x 1 columns]

In [117]:
scaler = preprocessing.MinMaxScaler()
player_df_sinNulos[["crossing"]] = scaler.fit_transform(player_df_sinNulos[["crossing"]])
player_df_sinNulos[["crossing"]]

crossing
0      0.511036
1      0.777231
2      0.745202
3      0.614443
4      0.468810
5      0.814419
6      0.498680
7      0.864203
8      0.071977
9      0.694458
10     0.239923
11     0.726607
12     0.503839
13     0.642514
14     0.802303
15     0.580374
16     0.643114
17     0.496161
18     0.761756
19     0.419866
20     0.770274
21     0.491843
22     0.661468
23     0.689779
24     0.515835
26     0.515835
27     0.431862
28     0.563820
29     0.704055
32     0.424664
...         ...
11030  0.563820
11031  0.767754
11032  0.727087
11033  0.585773
11034  0.620202
11035  0.647793
11036  0.697097
11037  0.514995
11038  0.731766
11039  0.419866
11040  0.814060
11041  0.796305
11042  0.753359
11043  0.617802
11044  0.687500
11045  0.812620
11046  0.704295
11047  0.910749
11048  0.699376
11049  0.622121
11050  0.263916
11051  0.371881
11052  0.755758
11053  0.707774
11054  0.498081
11055  0.431862
11056  0.734765
11057  0.685461
11058  0.488844
11059  0.866123

[10582 rows x 1 columns]

### Z-score

In [116]:
from scipy import stats
player_df['short_passing'] = stats.zscore(player_df['short_passing'])
player_df['short_passing']

0        0.017238
1        0.140868
2        0.351853
3        0.320761
4        0.325202
5        1.324605
6        0.237107
7        1.177285
8       -2.766282
9        0.635387
10       1.462300
11       0.760498
12      -0.175239
13      -0.164135
14       1.651816
15      -0.332182
16       0.108295
17      -0.316636
18       0.972223
19      -0.338105
20       0.565059
21      -0.249269
22       0.110516
23       0.428104
24      -0.323299
25       1.009238
26      -0.471358
27       0.268940
28      -0.273699
29       0.314838
           ...   
11030    0.713118
11031    0.734587
11032    0.421441
11033   -0.253711
11034   -0.745269
11035   -0.073078
11036    0.680545
11037   -0.550570
11038    0.639089
11039    0.446611
11040    1.273524
11041    1.767303
11042    0.531745
11043    0.200832
11044    0.274862
11045    1.235028
11046    0.237107
11047    1.214300
11048    0.650193
11049    0.406635
11050   -0.915537
11051   -0.619418
11052    1.183208
11053    0.565059
11054   -0

# 6. Codificar variables

> Las variables categóricas deben ser etiquetadas como variables numéricas, no como cadenas.

Codificar la variable `country_name` del Dataframe `match_df`

In [120]:
from sklearn import preprocessing

match_df_codificados = match_df

le = preprocessing.LabelEncoder()
le.fit(match_df_codificados['country_name'])

dict(zip(le.classes_, le.transform(le.classes_)))

{'Belgium': 0,
 'England': 1,
 'France': 2,
 'Germany': 3,
 'Italy': 4,
 'Netherlands': 5,
 'Poland': 6,
 'Portugal': 7,
 'Scotland': 8,
 'Spain': 9,
 'Switzerland': 10}

In [121]:
match_df_codificados['country_name_codificado'] = le.transform(match_df_codificados['country_name']) 
match_df_codificados

country_name               league_name     season  stage  \
0          Belgium    Belgium Jupiler League  2008/2009      1   
1          Belgium    Belgium Jupiler League  2008/2009      1   
2          Belgium    Belgium Jupiler League  2008/2009      1   
3          Belgium    Belgium Jupiler League  2008/2009      1   
4          Belgium    Belgium Jupiler League  2008/2009      1   
5          Belgium    Belgium Jupiler League  2008/2009      1   
6          Belgium    Belgium Jupiler League  2008/2009      1   
7          Belgium    Belgium Jupiler League  2008/2009      1   
8          Belgium    Belgium Jupiler League  2008/2009      1   
9          Belgium    Belgium Jupiler League  2008/2009     10   
10         Belgium    Belgium Jupiler League  2008/2009     10   
11         Belgium    Belgium Jupiler League  2008/2009     10   
12         Belgium    Belgium Jupiler League  2008/2009     10   
13         Belgium    Belgium Jupiler League  2008/2009     10   
14         Belgium    Belgium Jupiler League  2008/2009     10   
15         Belgium    Belgium Jupiler League  2008/2009     10   
16         Belgium    Belgium Jupiler League  2008/2009     10   
17         Belgium    Belgium Jupiler League  2008/2009     10   
18         Belgium    Belgium Jupiler League  2008/2009     11   
19         Belgium    Belgium Jupiler League  2008/2009     11   
20         Belgium    Belgium Jupiler League  2008/2009     11   
21         Belgium    Belgium Jupiler League  2008/2009     11   
22         Belgium    Belgium Jupiler League  2008/2009     11   
23         Belgium    Belgium Jupiler League  2008/2009     11   
24         Belgium    Belgium Jupiler League  2008/2009     11   
25         Belgium    Belgium Jupiler League  2008/2009     11   
26         Belgium    Belgium Jupiler League  2008/2009     11   
27         Belgium    Belgium Jupiler League  2008/2009     12   
28         Belgium    Belgium Jupiler League  2008/2009     12   
29         Belgium    Belgium Jupiler League  2008/2009     12   
...            ...                       ...        ...    ...   
25949  Switzerland  Switzerland Super League  2015/2016      4   
25950  Switzerland  Switzerland Super League  2015/2016      4   
25951  Switzerland  Switzerland Super League  2015/2016      4   
25952  Switzerland  Switzerland Super League  2015/2016      4   
25953  Switzerland  Switzerland Super League  2015/2016      4   
25954  Switzerland  Switzerland Super League  2015/2016      5   
25955  Switzerland  Switzerland Super League  2015/2016      5   
25956  Switzerland  Switzerland Super League  2015/2016      5   
25957  Switzerland  Switzerland Super League  2015/2016      5   
25958  Switzerland  Switzerland Super League  2015/2016      5   
25959  Switzerland  Switzerland Super League  2015/2016      6   
25960  Switzerland  Switzerland Super League  2015/2016      6   
25961  Switzerland  Switzerland Super League  2015/2016      6   
25962  Switzerland  Switzerland Super League  2015/2016      6   
25963  Switzerland  Switzerland Super League  2015/2016      6   
25964  Switzerland  Switzerland Super League  2015/2016      7   
25965  Switzerland  Switzerland Super League  2015/2016      7   
25966  Switzerland  Switzerland Super League  2015/2016      7   
25967  Switzerland  Switzerland Super League  2015/2016      7   
25968  Switzerland  Switzerland Super League  2015/2016      7   
25969  Switzerland  Switzerland Super League  2015/2016      8   
25970  Switzerland  Switzerland Super League  2015/2016      8   
25971  Switzerland  Switzerland Super League  2015/2016      8   
25972  Switzerland  Switzerland Super League  2015/2016      8   
25973  Switzerland  Switzerland Super League  2015/2016      8   
25974  Switzerland  Switzerland Super League  2015/2016      9   
25975  Switzerland  Switzerland Super League  2015/2016      9   
25976  Switzerland  Switzerland Super League  2015/2016      9   
25977  Switzerland  Switzerland Super League  

More about preprocessing in:
* https://scikit-learn.org/stable/modules/classes.html#module-sklearn.preprocessing